### PR TITLE
More threading options

### PR DESCRIPTION
**Pull request by samsartor**
This PR:
- Makes `par-map` a default optional dependency.
- Adds `rayon` as a non-default optional dependency.
- Adds `rayon_iter` to `OsmPbfReader` for iterating over `OsmObj`s.
- Adds `par_iter_with_nb_threads` to `OsmPbfReader`  in order to provide an explicit thread-count when using `par-map`.

The reasons for adding optional `rayon` integration are:
- It makes it easy for users of this  crate to read *and process*  `OsmObjs`s in parallel.
- It grants greater control over  thread configuration through `rayon`'s shared thread pool.
- It eliminates dependencies if the user is already importing `rayon` but not `par-map`.

### CODE REVIEW

**Comment by TeXitoi** [comment]
You also need to fmt on last stable.

**Comment by samsartor** [comment]
> What about memory consumption? Does bridge guarantee that memory will not grow too much?

Yep! Bridge is lazy. Looking at the source, it calls `.next()` only when the internal buffer has been emptied.

I saw the indentation screw up too. Fmt-ed commit incoming...

File: src/reader.rs [hunk 1]
```diff
@@ -66,7 +72,29 @@ impl<R: io::Read> OsmPbfReader<R> {
...
+
+    #[cfg(feature="par-map")]
+    /// Returns a parallel iterator on the OsmObj of the pbf file, using the
+    /// specified number of threads. See `par_iter`.
+    pub fn par_iter_with_nb_threads<'a>(&'a mut self, nb: usize) -> ParIter<'a, R> {
+        ParIter(self.blobs().with_nb_threads(nb).par_flat_map(result_blob_into_iter))
+    }
+
+    #[cfg(feature="rayon")]
+    /// Returns a rayon parallel iterator on the OsmObj of the pbf file.
```
[the remark this answers is not in the document]
Comment (samsartor) [reply src/reader.rs]: Good call.

File: src/reader.rs [hunk 2]
```diff
@@ -66,7 +72,29 @@ impl<R: io::Read> OsmPbfReader<R> {
...
+    ///
+    /// #Example
+    ///
+    /// ```
+    /// # extern crate rayon;
+    /// # use rayon::prelude::*;
+    /// let mut pbf = osmpbfreader::OsmPbfReader::new(std::io::empty());
+    /// println!("{}", pbf.rayon_iter().count());
+    /// ```
+    pub fn rayon_iter(&mut self) -> impl ParallelIterator<Item=Result<OsmObj>> + '_ where R: Send {
```
[the remark this answers is not in the document]
Comment (samsartor) [reply src/reader.rs]: I'll look into it.

**Comment by PureW** [comment]
Any updates on this @samsartor ?

**Comment by samsartor** [comment]
Got busy with other things. I can finish this up sometime this week.

I stopped working on it because protobuf parsing is single-treaded and seems to bottleneck anything else. I might have just been doing it wrong.

**Comment by TeXitoi** [comment]
Strange, it looks like protobuf parsing is done in parallel. Else even `par_map` would be useless.

**Comment by samsartor** [comment]
That's what I thought but `next_blob` requires a mutable lock on `OsmPbfReader`. And my experiments showed it running faster with 1 or 2 threads than with 3 or more. It was back in December so I don't have the logs. But I'd love it if someone could figure out what was going on.

**Pull request closed without merging**